src/som/interpreter/Interpreter.java

```diff
@@ -46,185 +66,171 @@ public Interpreter(final Universe universe) {
   }
 
-  private void doSuperSend(final int bytecodeIndex) {
+  private void doSuperSend(final int bytecodeIndex, Frame frame) {
     // Handle the SUPER SEND bytecode
-    SSymbol signature = (SSymbol) getMethod().getConstant(bytecodeIndex);
+    SSymbol signature = (SSymbol) getMethod(frame).getConstant(bytecodeIndex);
 
     // Send the message
     // Lookup the invokable with the given signature
-    SClass holderSuper = (SClass) getMethod().getHolder().getSuperClass();
+    SClass holderSuper = (SClass) getMethod(frame).getHolder().getSuperClass();
     SInvokable invokable = holderSuper.lookupInvokable(signature);
 
     if (invokable != null) {
       // Invoke the invokable in the current frame
-      invokable.invoke(getFrame(), this);
+      invokable.invoke(frame, this);
     } else {
       // Compute the number of arguments
       int numberOfArguments = signature.getNumberOfSignatureArguments();
 
       // Compute the receiver
-      SAbstractObject receiver = getFrame().getStackElement(numberOfArguments - 1);
+      SAbstractObject receiver = frame.getStackElement(numberOfArguments - 1);
 
-      receiver.sendDoesNotUnderstand(signature, universe, this);
+      receiver.sendDoesNotUnderstand(signature, universe, this, frame);
     }
   }
 
-  private void doReturnLocal() {
+  private SAbstractObject doReturnLocal(Frame frame) {
     // Handle the RETURN LOCAL bytecode
-    SAbstractObject result = getFrame().pop();
-
-    // Pop the top frame and push the result
-    popFrameAndPushResult(result);
+    return frame.pop();
   }
 
-  private void doReturnNonLocal() {
+  private SAbstractObject doReturnNonLocal(Frame frame) throws ReturnException {
     // Handle the RETURN NON LOCAL bytecode
-    SAbstractObject result = getFrame().pop();
+    SAbstractObject result = frame.pop();
 
     // Compute the context for the non-local return
-    Frame context = getFrame().getOuterContext(universe.nilObject);
+    Frame context = frame.getOuterContext(universe.nilObject);
 
     // Make sure the block context is still on the stack
     if (!context.hasPreviousFrame(universe.nilObject)) {
       // Try to recover by sending 'escapedBlock:' to the sending object
       // this can get a bit nasty when using nested blocks. In this case
       // the "sender" will be the surrounding block and not the object
       // that actually sent the 'value' message.
-      SBlock block = (SBlock) getFrame().getArgument(0, 0);
+      SBlock block = (SBlock) frame.getArgument(0, 0);
       SAbstractObject sender =
-          getFrame().getPreviousFrame().getOuterContext(universe.nilObject).getArgument(0, 0);
-
-      // pop the frame of the currently executing block...
-      popFrame();
-
-      // pop old arguments from stack
-      SMethod method = getFrame().getMethod();
-      int numArgs = method.getNumberOfArguments();
-      for (int i = 0; i < numArgs; i += 1) {
-        getFrame().pop();
-      }
```

Comment:
I am probably not yet fully awake, but do you still need to pop these arguments from the frame?

Reply:
Hmm, I still need to have a look. I think it kind of works right now because the following `sendEscapedBlock` call will only fetch the arguments from the given method and will ignore the ones that have not been popped

Reply:
There might just not be a test that runs out of stack.
I'd be careful with this.